Common/src/QuasarThreadPool.cpp

```diff
@@ -0,0 +1,96 @@
+/* © Copyright CERN, 2018.  All rights not expressly granted are reserved.
+ * QuasarThreadPool.cpp
+ *
+ *  Created on: 4 May 2018
+ *      Author: Piotr Nikiel <piotr@nikiel.info>
+ *
+ *  This file is part of Quasar.
+ *
+ *  Quasar is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public Licence as published by
+ *  the Free Software Foundation, either version 3 of the Licence.
+ *
+ *  Quasar is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public Licence for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with Quasar.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <algorithm>
+
+#include <QuasarThreadPool.h>
+#include <LogIt.h>
+
+namespace Quasar
+{
+
+ThreadPool::ThreadPool (unsigned int maxThreads, unsigned int maxJobs):
+        m_quit(false),
+        m_maxJobs(maxJobs)
+{
+    m_workers.reserve(maxThreads);
+    for (unsigned int i=0; i<maxThreads; ++i)
+        m_workers.emplace_back( [this](){this->work();} );
+}
+
+ThreadPool::~ThreadPool ()
+{
+    LOG(Log::TRC) << "Stopping threadpool";
+    m_quit = true;
+    m_conditionVariable.notify_all();
+    std::for_each( m_workers.begin(), m_workers.end(), [](std::thread &t){t.join();} );
```

Comment:
A timeout on the join? That way if some jobs do not shut down properly then at least we know which ones they are (if we have some way of ID-ing the jobs as per previous toString()/onError() comment).

Reply:
I have double thoughts here because of:
1) it is never clear what is the right timeout (I wouldn't like to add this to configuration unless necessary),
2) killing threads brutally might cause more mess (e.g. segfauls, core dumps) than being patient and letting the end user do that using "kill" tool if really necessary.

Reply:
Yep, I always seem to end up at roughly the same point when dealing with threads: How to handle the badly behaved instances that don't stop when they're supposed to? How long to wait etc.

To add some precision to my statement... I hadn't actually meant we should do anything brutal to lingering threads, rather that we should just log the threads that take a long time to stop - the sort of information that could help a developer improve their product. 

How long is a long time? Should it be configurable? I'd say if a thread hasn't stopped in 5s that's a long time: worth a warn/info entry in the logfile anyway.

Reply:
I like your idea. But it seems std::thread is a bit behind boost::thread in the sense that it doesn't yet support timed_join. I will try to see what can be done with std::thread::joinable().

Reply:
Hi @ben-farnham, I failed with std::thread. 
One could do with boost::thread but then better to go for thread_group because timed_join makes sense only on the whole group (to execute join concurrently rather than one after another).
So what I will merge is sth that prints a message in INF log level that the threadpool is being stopped and that it might require time, and a corresponding message that it's stopped.

There clearly is some room for user satisfaction improvement, so let's keep your idea this way.